### PR TITLE
createMemberInVO API method

### PR DESCRIPTION
**Pull request by pavolpluta**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code or performance improvement
- [x] Documentation changes or improvements

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My code has tests
- [x] My code has been tested manually
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/java/cz/muni/ics/perunproxyapi/application/facade/impl/RelyingPartyFacadeImpl.java [hunk 1]
```diff
@@ -119,4 +129,36 @@ public boolean hasAccessToService(@NonNull String rpIdentifier, @NonNull String
...
+        BidiMap<String, String> candidateAttrsMapper = new DualHashBidiMap<>(FacadeUtils.externalToInternalAttributeMapper(CANDIDATE_ATTR_MAPPER, CREATE_MEMBER_IN_VO, options));
+        if (!requestAttributes.keySet().containsAll(requiredAttrs)) {
+            throw new InvalidRequestParameterException("Not all required attributes have been sent in a request body.");
+        }
+
+        String login = null;
+        for (String loginAttr: loginAttrs) {
+            // asText() should NOT be used as it will return "null" instead of null.
+            // textValue() returns null for any non-String values and empty String.
+            if (requestAttributes.get(loginAttr).textValue() != null) {
```
Comment (dBucik) [inline src/main/java/cz/muni/ics/perunproxyapi/application/facade/impl/RelyingPartyFacadeImpl.java:150]: Check for a nonempty string might be a good idea as well.

File: src/main/java/cz/muni/ics/perunproxyapi/application/service/RelyingPartyService.java [hunk 2]
```diff
@@ -73,4 +80,23 @@ boolean hasAccessToService(@NonNull DataAdapter adapter,
...
+     * @param login Login which will be used for UserExtSource.
+     * @param requestAttributes Attributes from request body.
+     * @param perunAttributesMapper Map where key is the external attr name and value internal, Perun attr identifier.
+     * @param candidateAttributesMapper Map where key is the candidate attr name used in Perun and value is external attr name.
+     * @return TRUE if member was created, otherwise FALSE>
+     * @throws PerunUnknownException Thrown as wrapper of unknown exception thrown by Perun interface.
+     * @throws PerunConnectionException Thrown when problem with connection to Perun interface occurs.
+     * @throws InvalidRequestParameterException Invalid parameter given in the request body.
+     */
+    boolean createMemberInVo(@NonNull FullAdapter rpcAdapter, @NonNull String extSourceIdentifier, @NonNull Long voId,
```
Comment (dBucik) [inline src/main/java/cz/muni/ics/perunproxyapi/application/service/RelyingPartyService.java:98]: This method is about the user, not about the relying party. Please move it to the correct class.

File: src/main/java/cz/muni/ics/perunproxyapi/application/service/ServiceUtils.java [hunk 3]
```diff
@@ -64,4 +75,59 @@ public static String wrapCapabilityToAARC(@NonNull String capability, @NonNull S
...
+     *
+     * @param attributes Attributes from request body.
+     * @param perunAttributesMapper Map where key is the external attr name and value internal, Perun attr identifier.
+     * @param candidateAttributesMapper Map where key is the candidate attr name used in Perun and value is external attr name.
+     * @return Candidate object with attributes, without UserExtSoruce.
+     */
+    public static Candidate createCandidateWithoutUserExtSource(Map<String, JsonNode> attributes, Map<String, String> perunAttributesMapper, BidiMap<String, String> candidateAttributesMapper) {
+        Candidate candidate = new Candidate();
+
+        attributes.forEach((attrName, attrValue) -> {
```
Comment (dBucik) [inline src/main/java/cz/muni/ics/perunproxyapi/application/service/ServiceUtils.java:89]: Some normalization of the attr names might be a good idea. I.e., we could use only LOWERCASE names.

File: src/main/java/cz/muni/ics/perunproxyapi/application/service/ServiceUtils.java [hunk 4]
```diff
@@ -64,4 +75,59 @@ public static String wrapCapabilityToAARC(@NonNull String capability, @NonNull S
...
+            if (candidateAttributesMapper.containsValue(attrName)) {
+                mapCandidateAttribute(candidate, candidateAttributesMapper.getKey(attrName), attrValue.asText());
+            } else {
+                if (!perunAttributesMapper.containsKey(attrName)) {
+                    throw new ConfigurationException(String.format("Configuration file does not provide mapping for attribute '%s'", attrName));
```
Comment (dBucik) [inline src/main/java/cz/muni/ics/perunproxyapi/application/service/ServiceUtils.java:94]: I think it is enough to log this fact and continue with the execution instead.

**Review by dBucik — CHANGES_REQUESTED** [review]
Just small changes needed, otherwise, it seems quite OK. Have you been also testing it manually?

File: src/main/java/cz/muni/ics/perunproxyapi/application/service/ServiceUtils.java (hunk 3 above)
Comment (pavolpluta) [reply src/main/java/cz/muni/ics/perunproxyapi/application/service/ServiceUtils.java:89]: Attr names from request equals to attr values in configuration -> candidate_attr_mapper. Do we really want to make changes programmatically? Wouldn't it be sufficient to just set attr values in the configuration properly?

File: src/test/resources/configs/proxyuser-facade.yml [hunk 5]
```diff
@@ -10,10 +10,27 @@
...
+    vo_id: 1
+    required_attributes: ["cn", "eduPersonPrincipalName", "eduPersonScopedAffiliation"] # ATTRIBUTES THAT MUST BE PRESENT IN THE REQUEST BODY, PART ATTRIBUTES
+    login_attributes: ["eduPersonTargetedId", "userId"] # ATTRIBUTES THAT WILL BE USED FOR LOGIN, FIRST NONEMPTY FROM LIST IS USED (i.e. in case that EPTID is empty in request body, userId is used)
+    attr_mapper: # MAPPING OF INTERNAL PERUN IDENTIFIERS TO EXTERNAL NAMES
+      eduPersonPrincipalName: "urn:perun:ues:attribute-def:def:eppn"
+      sn: "urn:perun:ues:attribute-def:def:sn"
+      eduPersonTargetedId: "urn:perun:ues:attribute-def:def:eptid"
+      eduPersonScopedAffiliation: "urn:perun:ues:attribute-def:def:affiliation"
+    candidate_attr_mapper: # MAPPING OF CANDIDATE PROPERTY NAME TO EXTERNAL NAME
+      firstName: "cn"
```
Comment (vyskocilpavel) [inline src/test/resources/configs/proxyuser-facade.yml:34]: cn is commonName.

Should be givenName on this line

File: src/main/java/cz/muni/ics/perunproxyapi/presentation/rest/controllers/ProxyUserProtectedController.java [hunk 6]
```diff
@@ -466,4 +462,47 @@ public JsonNode ga4ghByLogin(@PathVariable(value = LOGIN) String login)
         return facade.ga4ghByLogin(login);
     }
 
+    /**
+     * Create new member in the VO.
+     *
+     * EXAMPLE CURL:
+     * curl --request POST \
+     *   --url http://localhost:8081/papi/auth/relying-party/createMemberInVo \
```
Comment (vyskocilpavel) [inline src/main/java/cz/muni/ics/perunproxyapi/presentation/rest/controllers/ProxyUserProtectedController.java:470]: bad example URL

**Pull request merged**